Use numpy 2.1.1 and remove oldest-supported-numpy which is now deprec…

ccoulombe opened the pull request:
…ated

`oldest-supported-numpy` is now deprecated. This means we'll need to build with numpy 2.0 and backward compat 1.x.
https://github.com/scipy/oldest-supported-numpy/commit/23ece5a161d5957f5142991393d2ae689ee45b9d (edited)

ccoulombe committed:
Use numpy 2.1.1 and remove oldest-supported-numpy which is now deprecated

ccoulombe commented:
Closes https://github.com/ComputeCanada/software-stack/issues/156

mboisson commented on build_wheel.sh:
```diff
@@ -100,8 +99,12 @@ PATCHES=""
 
-PYTHON_DEPS_DEFAULT=""
-MODULE_BUILD_DEPS_DEFAULT="oldest-supported-numpy/$OLDEST_SUPPORTED_NUMPY_VERSION python-build-bundle pytest cython/$CYTHON_VERSION"
+PYTHON_DEPS_DEFAULT="numpy==2.1.1"
```
devrait être le module ?

ccoulombe replied:
Ce sera équivalent. J'avais fait cette PR avant que l'on discute de créer un module pour 2.1.1.

mboisson reviewed (changes requested):
Ligne 8 à enlever
 adjust_numpy_requirements_based_on_link_info à modifier/enlever pour ne plus mettre la version minimum de numpy si numpy_build_version >= 2

ccoulombe committed:
Use numpy module instead of wheel

ccoulombe committed:
Updated adjust_numpy_requirements_based_on_link_info for Numpy 2.0

Fixed grepping the version

Do not pin minimal numpy when 2.x was used

Pin minimal but also set lower than when numpy 1.x was used

mboisson commented on build_wheel.sh:
```diff
@@ -492,14 +492,22 @@ function adjust_numpy_requirements_based_on_link_info()
 		rm -rf $tmpdir
 		if [[ $num_links -gt 0 ]]; then
-			numpy_build_version=$(pip show numpy | grep Version | awk '{print $2}' | sed -e "s/\([0-9]\.[0-9]*\)\..*/\1/g")
+			numpy_build_version=$(pip show numpy | grep -E '^Version:' | awk '{print $2}' | sed -e "s/\([0-9]\.[0-9]*\)\..*/\1/g")
```
```
python -c "import numpy; print(numpy.__version__)"
2.1.1
```

?

ccoulombe replied:
Yup, way better and clearer :
```
python -c "import numpy; print(numpy.__version__)" | grep -P -oe '\d+.\d'
```

mboisson commented:
minor detail, otherwise LGTM.

ccoulombe committed:
Simplify numpy version getting and parsing
